### PR TITLE
4.1.x - user sees the upload status of multiple (in progress) videos while they continue recording

### PR DESCRIPTION
**Pull request by aaronshiel**

### CODE REVIEW

File: client/src/components/nav-bar.tsx [hunk 1]
```diff
@@ -206,22 +215,42 @@ function NavMenu(props: {
...
+      upload.uploadStatus !== UploadStatus.CANCELLED
+  ).length;
+  const numUploadsComplete = uploads?.filter(
+    (upload) => upload.uploadStatus == UploadStatus.DONE
+  ).length;
 
+  const [isDrawerOpen, setIsDrawerOpen] = React.useState(false);
+  //if there are no uploads, defaults to true, else if there are any uploads that aren't yet cancelled, should not be disabled
+  const disableUploadsButton = uploads
+    ? uploads.find(
```
Comment (beatthat) [inline client/src/components/nav-bar.tsx:247]: use falsyness...almost never want to check for `== undefined`

```typescript
? Boolean(uploads.find(...))
```

```

File: client/src/components/record/video-player.tsx [hunk 2]
```diff
@@ -106,46 +96,72 @@ function VideoPlayer(props: {
...
         }}
       >
         <Typography
           data-cy="warning"
           style={{
             textAlign: "center",
             visibility:
-              curAnswer.videoSrc &&
-              videoLength < (curAnswer.minVideoLength || 0)
+              recordState.curAnswer!.videoSrc &&
```
Comment (beatthat) [inline client/src/components/record/video-player.tsx:118]: let's not use the `not-null` assertion anywhere, most lint checkers consider this an error

File: client/src/hooks/graphql/use-with-upload-status.tsx [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+          }
+          data.forEach((u) => {
+            const findUpload = uploads.find(
+              (up) => up.question._id === u.question._id
+            );
+            // add  || findUpload.uploadStatus == UploadStatus.UPLOAD_IN_PROGRESS   if you are cypress testing uploadProgress
+            if (
+              !findUpload ||
+              findUpload.uploadStatus !== u.uploadStatus ||
+              findUpload.uploadStatus == UploadStatus.UPLOAD_IN_PROGRESS
```
Comment (beatthat) [inline client/src/hooks/graphql/use-with-upload-status.tsx:87]: ===

File: client/src/hooks/graphql/use-with-upload-status.tsx [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+      task.uploadStatus === UploadStatus.CANCELLED ||
+      task.uploadStatus === UploadStatus.TRANSCRIBE_FAILED ||
+      task.uploadStatus === UploadStatus.UPLOAD_FAILED ||
+      task.uploadStatus === UploadStatus.DONE
+    );
+  }
+
+  function isTaskPolling(task: UploadTask) {
+    return (
+      //task.uploadStatus !== UploadStatus.PENDING &&
```
Comment (beatthat) [inline client/src/hooks/graphql/use-with-upload-status.tsx:114]: delete commented code

File: client/src/pages/record.tsx [hunk 5]
```diff
@@ -128,10 +121,17 @@ function RecordPage(props: {
       navigate("/");
     }
   }
-
   function switchAnswer(onNav: () => void) {
-    if (curAnswer?.isEdited) {
-      setConfirmLeave({ callback: onNav });
+    if (curAnswer!.isEdited) {
+      if (curAnswer!.recordedVideo && !curAnswer!.isUploading) {
```
Comment (beatthat) [inline client/src/pages/record.tsx:126]: no not-null

File: cypress/cypress/support/functions.ts [hunk 6]
```diff
@@ -241,29 +241,4 @@ export function cyMockUpload(
...
-          errors: null,
-        },
-        headers: {
-          "Content-Type": "application/json",
-        },
-      })
-    );
-  });
-}
+}
```
Comment (beatthat) [inline cypress/cypress/support/functions.ts:244]: new line at EOF

**Review by beatthat — CHANGES_REQUESTED** [review]
looks great!

A few inline change requests, but the more spanning one is we should not use the non-null operator ever. I will look at if I can enforce that rule in lint

File: cypress/cypress/support/functions.ts (hunk 6 above)
Comment (aaronshiel) [reply cypress/cypress/support/functions.ts]: fd9f53b

File: client/src/pages/record.tsx (hunk 5 above)
Comment (aaronshiel) [reply client/src/pages/record.tsx]: fd9f53b

File: client/src/hooks/graphql/use-with-upload-status.tsx (hunk 4 above)
Comment (aaronshiel) [reply client/src/hooks/graphql/use-with-upload-status.tsx]: fd9f53b

File: client/src/hooks/graphql/use-with-upload-status.tsx (hunk 3 above)
Comment (aaronshiel) [reply client/src/hooks/graphql/use-with-upload-status.tsx:86]: fd9f53b

File: client/src/components/record/video-player.tsx (hunk 2 above)
Comment (aaronshiel) [reply client/src/components/record/video-player.tsx]: fd9f53b

File: client/src/components/nav-bar.tsx (hunk 1 above)
Comment (aaronshiel) [reply client/src/components/nav-bar.tsx]: fd9f53b

**Review by beatthat — APPROVED** [review]
looks great

**Pull request merged**